File: frontend/js/views/kiddie_tax.js
// Kiddie Tax § 1(g) Calculator.
// Unearned income above $2,600 (2024 / $2,700 2025) taxed at PARENT'S
// marginal rate. Applies to: under-18, OR full-time student 19-23 whose
// earned income < 50% of support. Defeats the "give appreciated stock
// to kid to sell at 0% LT cap-gains rate" strategy unless kid is grown.

import { api } from '../api.js';
import { esc } from '../util.js';
import { t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';

let state = {
    year: new Date().getFullYear(),
    kid_age: 12,
    kid_earned_income: 0,
    kid_unearned_income: 8_000,
    parent_marginal_rate: 0.32,
    parent_lt_cap_gains_rate: 0.20,
};

export async function renderKiddieTax(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.kiddie.h1.title">// KIDDIE TAX § 1(g)</span></h1>
        <p class="muted small" data-i18n="view.kiddie.hint.intro">
            Unearned income &gt; $2,600 (2024) taxed at <strong>parent's marginal rate</strong>.
            Applies to: under-18, OR full-time student 19-23 whose earned income &lt; 50%
            of support. Defeats the classic "gift appreciated stock to kid" strategy
            unless kid is past age threshold or has substantial earned income.
        </p>
        <div class="chart-panel">
            <h2 data-i18n="view.kiddie.h2.inputs">Inputs</h2>
            <form id="kt-form" class="inline-form">
                <label><span data-i18n="view.kiddie.label.year">Year</span>
                    <input type="number" step="1" name="year" value="${state.year}"></label>
                <label><span data-i18n="view.kiddie.label.kid_age">Kid age</span>
                    <input type="number" step="1" name="kid_age" value="${state.kid_age}" min="0" max="25"></label>
                <label><span data-i18n="view.kiddie.label.kid_earned">Kid earned income ($)</span>
                    <input type="number" step="0.01" name="kid_earned_income" value="${state.kid_earned_income}"></label>
                <label><span data-i18n="view.kiddie.label.kid_unearned">Kid unearned income ($)</span>
                    <input type="number" step="0.01" name="kid_unearned_income" value="${state.kid_unearned_income}"></label>
                <label><span data-i18n="view.kiddie.label.parent_rate">Parent's marginal rate %</span>
                    <input type="number" step="0.5" name="parent_marginal_rate" value="${(state.parent_marginal_rate * 100).toFixed(1)}"></label>
                <label><span data-i18n="view.kiddie.label.parent_lt_rate">Parent's LT cap-gains rate %</span>
                    <input type="number" step="0.5" name="parent_lt_cap_gains_rate" value="${(state.parent_lt_cap_gains_rate * 100).toFixed(1)}"></label>
                <button class="primary" type="submit" data-i18n="view.kiddie.btn.compute">Compute</button>
            </form>
        </div>
        <div id="kt-output"></div>
    `;
    document.getElementById('kt-form').addEventListener('submit', (e) => {
        e.preventDefault();
        const fd = new FormData(e.target);
        state.year = Number(fd.get('year'));
        state.kid_age = Number(fd.get('kid_age'));
        state.kid_earned_income = Number(fd.get('kid_earned_income')) || 0;
        state.kid_unearned_income = Number(fd.get('kid_unearned_income')) || 0;
        state.parent_marginal_rate = (Number(fd.get('parent_marginal_rate')) || 32) / 100;
        state.parent_lt_cap_gains_rate = (Number(fd.get('parent_lt_cap_gains_rate')) || 20) / 100;
        renderOutput();
    });
    renderOutput();
}

async function renderOutput() {
    const el = document.getElementById('kt-output');
    if (!el) return;
    let r;
    try {
        r = await api.calcKiddieTax({
            year: state.year,
            kid_age: state.kid_age,
            kid_earned_income_usd: state.kid_earned_income,
            kid_unearned_income_usd: state.kid_unearned_income,
            parent_marginal_rate_fraction: state.parent_marginal_rate,
            parent_lt_cap_gains_fraction: state.parent_lt_cap_gains_rate,
        });
    } catch (e) {
        el.innerHTML = `<p class="neg">${esc(String(e))}</p>`;
        return;
    }
    const kiddieTaxApplies = r.kiddie_tax_applies;
    const effStdDed = r.eff_std_deduction_usd;
    const offsetByStdDed = r.offset_by_std_ded_usd;
    const kidsOwnRateAmount = r.kids_rate_amount_usd;
    const subjectToParentRate = r.subject_to_parent_rate_usd;
    const kidsTaxOnEarned = r.tax_on_earned_usd;
    const kidsTaxOnOwnUnearned = r.tax_at_kids_rate_usd;
    const parentRateTax = r.tax_at_parent_rate_usd;
    const totalKidTax = r.total_kid_tax_usd;
    const parentDirectTax = r.parent_direct_tax_usd;
    const savingsVsParent = r.savings_vs_parent_usd;

    el.innerHTML = `
        <div class="chart-panel ${kiddieTaxApplies ? 'neg' : 'pos'}">
            <h2 data-i18n="view.kiddie.h2.applies">Kiddie tax applies?</h2>
            <p style="font-size:1.4em">
                <strong>${kiddieTaxApplies ? esc(t('view.kiddie.status.yes')) : esc(t('view.kiddie.status.no'))}</strong>
            </p>
            <p class="muted small">
                ${kiddieTaxApplies
                    ? esc(t('view.kiddie.status.applies_reason'))
                    : esc(t('view.kiddie.status.exempt_reason'))}
            </p>
        </div>
        <div class="chart-panel">
            <h2 data-i18n="view.kiddie.h2.breakdown">Tax breakdown</h2>
            <table class="trades">
                <tbody>
                    <tr><td data-i18n="view.kiddie.row.std_ded">Standard deduction (dependent)</td>
                        <td>$${effStdDed.toLocaleString()}</td></tr>
                    <tr><td data-i18n="view.kiddie.row.offset_by_std_ded">Unearned offset by std deduction</td>
                        <td>$${offsetByStdDed.toLocaleString()}</td></tr>
                    <tr><td data-i18n="view.kiddie.row.kids_rate">Unearned at kid's rate (10%)</td>
                        <td>$${kidsOwnRateAmount.toLocaleString()}</td></tr>
                    <tr><td data-i18n="view.kiddie.row.parent_rate_amt">Unearned at parent's rate</td>
                        <td>$${subjectToParentRate.toLocaleString()}</td></tr>
                    <tr><td data-i18n="view.kiddie.row.tax_earned">Tax on earned</td>
                        <td class="neg">$${kidsTaxOnEarned.toFixed(2)}</td></tr>
                    <tr><td data-i18n="view.kiddie.row.tax_own">Tax at kid's rate (10%)</td>
                        <td class="neg">$${kidsTaxOnOwnUnearned.toFixed(2)}</td></tr>
                    <tr><td data-i18n="view.kiddie.row.tax_parent">Tax at parent's rate</td>
                        <td class="neg">$${parentRateTax.toFixed(2)}</td></tr>
                    <tr><td><strong data-i18n="view.kiddie.row.total">Total tax</strong></td>
                        <td><strong class="neg">$${totalKidTax.toFixed(2)}</strong></td></tr>
                </tbody>
            </table>
        </div>
        <div class="chart-panel">
            <h2 data-i18n="view.kiddie.h2.comparison">Vs. parent holding the asset</h2>
            <div class="cards">
                <div class="card">
                    <div class="label" data-i18n="view.kiddie.card.parent_direct">Parent's tax @ LT cap-gains</div>
                    <div class="value">$${parentDirectTax.toFixed(2)}</div>
                </div>
                <div class="card">
                    <div class="label" data-i18n="view.kiddie.card.kids_total">Kid's total tax</div>
                    <div class="value">$${totalKidTax.toFixed(2)}</div>
                </div>
                <div class="card ${savingsVsParent > 0 ? 'pos' : 'neg'}">
                    <div class="label" data-i18n="view.kiddie.card.savings">Gift-to-kid savings</div>
                    <div class="value">$${savingsVsParent.toFixed(2)}</div>
                </div>
            </div>
            <p class="muted small" data-i18n="view.kiddie.strategy">
                Strategy: gift basis-stepped stock to kids ONLY if they're 18+ AND have
                low income (use 0% LT cap-gains bracket up to $47,025 in 2024). Under 18:
                limit gifts to $2,600/yr of unearned income to stay within kid-rate tier.
                529 plans bypass kiddie tax entirely (growth is tax-free).
            </p>
        </div>
    `;
}

File: crates/traderview-core/src/kiddie_tax.rs
//! Kiddie tax (§ 1(g)) calculator. A dependent child's unearned income is taxed
//! in tiers: the first half of the threshold ($1,300 for 2024 / $1,350 for 2025)
//! is offset by the dependent standard deduction, the next half is taxed at the
//! child's own rate (10%), and anything above the threshold is taxed at the
//! parent's marginal rate. The rule applies under age 18, or to full-time
//! students 19–23. Compares the child's total tax against the parent simply
//! holding the asset and paying long-term cap-gains — the "gift appreciated
//! stock to the kid" strategy only wins outside the kiddie-tax ages. Faithful
//! port of the former client-side calculator. Pure compute, not advice.

use serde::{Deserialize, Serialize};

/// Child's single-filer brackets (cap, rate) as used by the original.
const KID_BRACKETS_SINGLE: [(f64, f64); 5] = [
    (11_600.0, 0.10),
    (47_150.0, 0.12),
    (100_525.0, 0.22),
    (191_950.0, 0.24),
    (f64::INFINITY, 0.37),
];

#[derive(Debug, Clone, Deserialize)]
pub struct KiddieTaxInput {
    pub year: u32,
    pub kid_age: u32,
    pub kid_earned_income_usd: f64,
    pub kid_unearned_income_usd: f64,
    /// Parent's marginal rate as a fraction (0.32 = 32%).
    pub parent_marginal_rate_fraction: f64,
    /// Parent's long-term cap-gains rate as a fraction.
    pub parent_lt_cap_gains_fraction: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq, Default)]
pub struct KiddieTaxReport {
    pub kiddie_tax_applies: bool,
    pub eff_std_deduction_usd: f64,
    pub offset_by_std_ded_usd: f64,
    pub kids_rate_amount_usd: f64,
    pub subject_to_parent_rate_usd: f64,
    pub tax_on_earned_usd: f64,
    pub tax_at_kids_rate_usd: f64,
    pub tax_at_parent_rate_usd: f64,
    pub total_kid_tax_usd: f64,
    pub parent_direct_tax_usd: f64,
    /// Parent-direct tax − child's total: positive means gifting to the kid wins.
    pub savings_vs_parent_usd: f64,
    pub valid: bool,
}

fn round2(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn kid_tax(taxable: f64) -> f64 {
    let mut owe = 0.0;
    let mut last_cap = 0.0;
    for &(cap, rate) in KID_BRACKETS_SINGLE.iter() {
        let slice = (taxable.min(cap) - last_cap).max(0.0);
        owe += slice * rate;
        if taxable <= cap {
            break;
        }
        last_cap = cap;
    }
    owe
}

pub fn generate(i: &KiddieTaxInput) -> KiddieTaxReport {
    let std_ded: f64 = if i.year >= 2025 { 1_350.0 } else { 1_300.0 };
    let tier: f64 = if i.year >= 2025 { 2_700.0 } else { 2_600.0 };
    let half_tier = tier / 2.0;

    let applies = i.kid_age < 18 || (19..=23).contains(&i.kid_age);

    // Dependent standard deduction: greater of the floor or earned + $450.
    let eff_std_ded = std_ded.max(i.kid_earned_income_usd + 450.0);

    let subject_to_parent = (i.kid_unearned_income_usd - tier).max(0.0);
    let kids_rate_amount = (i.kid_unearned_income_usd.min(tier) - half_tier).max(0.0);
    let offset_by_std_ded = i.kid_unearned_income_usd.min(half_tier);
    let tax_on_earned = kid_tax((i.kid_earned_income_usd - eff_std_ded).max(0.0));
    let tax_at_kids_rate = kids_rate_amount * 0.10;
    let tax_at_parent_rate = subject_to_parent * i.parent_marginal_rate_fraction;

    let total_kid_tax = if applies {
        tax_on_earned + tax_at_kids_rate + tax_at_parent_rate
    } else {
        tax_on_earned + kid_tax(i.kid_unearned_income_usd)
    };

    let parent_direct_tax = i.kid_unearned_income_usd * i.parent_lt_cap_gains_fraction;
    let savings_vs_parent = parent_direct_tax - total_kid_tax;

    KiddieTaxReport {
        kiddie_tax_applies: applies,
        eff_std_deduction_usd: round2(eff_std_ded),
        offset_by_std_ded_usd: round2(offset_by_std_ded),
        kids_rate_amount_usd: round2(kids_rate_amount),
        subject_to_parent_rate_usd: round2(subject_to_parent),
        tax_on_earned_usd: round2(tax_on_earned),
        tax_at_kids_rate_usd: round2(tax_at_kids_rate),
        tax_at_parent_rate_usd: round2(tax_at_parent_rate),
        total_kid_tax_usd: round2(total_kid_tax),
        parent_direct_tax_usd: round2(parent_direct_tax),
        savings_vs_parent_usd: round2(savings_vs_parent),
        valid: true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> KiddieTaxInput {
        KiddieTaxInput {
            year: 2026,
            kid_age: 12,
            kid_earned_income_usd: 0.0,
            kid_unearned_income_usd: 8_000.0,
            parent_marginal_rate_fraction: 0.32,
            parent_lt_cap_gains_fraction: 0.20,
        }
    }

    // Pins cross-checked against the JS compute() in Python.
    #[test]
    fn default_applies_2025_tier() {
        let d = generate(&base());
        assert!(d.valid);
        assert!(d.kiddie_tax_applies);
        assert!(close(d.eff_std_deduction_usd, 1_350.0));
        assert!(close(d.offset_by_std_ded_usd, 1_350.0));
        assert!(close(d.kids_rate_amount_usd, 1_350.0));
        assert!(close(d.subject_to_parent_rate_usd, 5_300.0));
        assert!(close(d.tax_on_earned_usd, 0.0));
        assert!(close(d.tax_at_kids_rate_usd, 135.0));
        assert!(close(d.tax_at_parent_rate_usd, 1_696.0));
        assert!(close(d.total_kid_tax_usd, 1_831.0));
        assert!(close(d.parent_direct_tax_usd, 1_600.0));
        assert!(close(d.savings_vs_parent_usd, -231.0));
    }

    #[test]
    fn exempt_adult_uses_kid_brackets_on_all() {
        let d = generate(&KiddieTaxInput { kid_age: 25, ..base() });
        assert!(!d.kiddie_tax_applies);
        // All $8,000 unearned at the kid's 10% bracket → $800.
        assert!(close(d.total_kid_tax_usd, 800.0));
    }

    #[test]
    fn year_2024_uses_lower_tier() {
        let d = generate(&KiddieTaxInput { year: 2024, ..base() });
        assert!(close(d.kids_rate_amount_usd, 1_300.0));
        assert!(close(d.subject_to_parent_rate_usd, 5_400.0));
        assert!(close(d.total_kid_tax_usd, 1_858.0));
    }

    #[test]
    fn student_19_to_23_applies() {
        let d = generate(&KiddieTaxInput { kid_age: 20, ..base() });
        assert!(d.kiddie_tax_applies);
    }
}
